Notifications - Rework data flow & add ability to locally page

tonisevener opened the pull request:
**Phabricator:** https://phabricator.wikimedia.org/T288662

### Notes
This PR modifies the data handling of Notification Center. It refactors out `NSFetchedResultsController` into basic `NSFetchRequests` in `RemoteNotificationsController`, and uses a diffable data source to animate the changes instead of `CollectionViewUpdater`. It also adds paging through local results and a basic edit mode to see the different cell styles.

I tested the performance and I didn't notice any issues with this particular code. I found other areas of improvement, but I will handle that in a followup PR.

### Test Steps
1. Fresh install app. Login to an account with notifications. Go to Settings > Languages and add additional app languages to get the number of notifications pretty large. You might want to consider tweaking the fetch limit [here](https://github.com/wikimedia/wikipedia-ios/blob/ba6d5225bbac52f533f30d221d86b1120e8a4920/WMF%20Framework/Remote%20Notifications/RemoteNotificationsController.swift#L31) to something lower if you have a smaller number of notifications so you can experience local paging.
2. Go to notification center. Wait for it to complete importing, then scroll around. Scroll down to the bottom, confirm next 50 cells are fetched and added to the list.
3. Tap the Edit button at the top, the view will switch to edit mode. Tap the circles in the cells and confirm they show as checked. Scroll down, then scroll back up. Confirm your checked cells are still checked.

### FYI
I've been working ahead in a `T288662-paging-future` prototype branch, if you're curious to see the data flow for future functionality. It has a pull to refresh, mark as read (just tapping the cell background for this), filtering (just one type), and 10,000 mock notifications importing for testing performance.

Dereck1980 reviewed (changes requested):
Oh yeah

tonisevener commented:
> @staykids Going to take over this one for a bit to clean up the diff. I might be able to remove the stacked dependencies.

@staykids Done, this is ready for review again.

staykids commented on Wikipedia/Code/NotificationsCenterViewModel.swift:
```diff
@@ -25,57 +37,95 @@ final class NotificationsCenterViewModel: NSObject {
+                return
+            }
+            
+            NotificationCenter.default.addObserver(self, selector: #selector(self.contextObjectsDidChange(_:)), name: Notification.Name.NSManagedObjectContextObjectsDidChange, object: self.remoteNotificationsController.viewContext)
```
Would it make sense for this observer subscription to just exist outside of this block, at the top of this function?

tonisevener replied:
@staykids actually this location was deliberate - in an earlier version I had this listener set up in the `init` statement of this view model even. The problem was as things are importing (particularly a large set of notifications), the `contextObjectsDidChange` handler starts firing and data is displayed before importing is completed. So the importing state goes away quickly and we start to see cells flying in and reordering for a while. Also because the main thread is so active inserting cells, it's practically unscrollable, so it's of little use to the user to see this data while it's importing. By only adding the listener here in this completion block, I can be sure languages are fully imported and the data is a bit more stable before `contextObjectsDidChange` fires.

staykids reviewed (changes requested):
Nice – feels great increasing our flexibility by reducing our dependence on `NSFetchedResultsController`. I had to reduce my fetch limit as you suggested because I didn't have enough notifications to test with.

Some comments inline below, also, here's a bug I noticed:

1. Have a sufficiently long list of notifications (more than a screenful)
2. Set the fetch limit to 5: `fetchNotifications(fetchLimit: Int = 5...`
3. Toggle editing mode on and start scrolling the list
4. Some of the incoming displayed notifications are not displayed in the edit mode state, but are instead displayed in the default display state (their leading image is the notification type instead of the checkmark selection view)

tonisevener commented on WMF Framework/Remote Notifications/RemoteNotificationsOperationsController.swift:
```diff
@@ -54,53 +55,79 @@ class RemoteNotificationsOperationsController: NSObject {
+        
+        guard !isLocked else {
+            assertionFailure("Failure setting up notifications core data stack.")
+            completion(.failureSettingUpModelController)
```
@staykids It's meant to be more generic, it's basically an error if there's any issue setting up the Core Data stack. I changed this name to `dataUnavailable`, hopefully that's a little better.

tonisevener commented on Wikipedia/Code/NotificationsCenterCell.swift:
```diff
@@ -2,6 +2,7 @@ import UIKit
 
 protocol NotificationsCenterCellDelegate: AnyObject {
     func userDidTapSecondaryActionForCellIdentifier(id: String)
+    func toggleCheckedStatus(viewModel: NotificationsCenterCellViewModel)
```
@staykids good idea, this is done!

tonisevener commented on Wikipedia/Code/NotificationsCenterCell.swift:
```diff
@@ -24,6 +26,12 @@ final class NotificationsCenterCell: UICollectionViewCell {
         return view
     }()
+    
+    lazy var leadingImageTapGestureRecognizer: UITapGestureRecognizer = {
```
@staykids I ended up exploring some of the the system-provided edit mode API and refactoring to that. I think it feels a little cleaner to not pass around a custom `editMode` flag all over the place, plus this made the whole cell tappable without having to add a hidden button overlay or something. Let me know what you think!

staykids replied:
👍🏽 feels like a positive change – nice to not have to manually juggle the Edit bar button item as well.

tonisevener commented on Wikipedia/Code/NotificationsCenterCell.swift:
```diff
@@ -376,4 +385,11 @@ final class NotificationsCenterCell: UICollectionViewCell {
         metaActionButton.setImage(image, for: .normal)
     }
 
+    @objc func tappedLeadingImage() {
```
@staykids I was able to remove this entirely with the edit mode refactor.

tonisevener commented on Wikipedia/Code/NotificationsCenterCellStyle.swift:
```diff
@@ -171,5 +171,9 @@ struct NotificationsCenterCellStyle {
         }
     }
+    
+    func isLeadingImageTapGestureEnabled(_ displayState: NotificationsCenterCellDisplayState) -> Bool {
```
@staykids I also removed this entirely.

tonisevener commented on Wikipedia/Code/NotificationsCenterCellViewModel.swift:
```diff
@@ -19,16 +19,81 @@ final class NotificationsCenterCellViewModel {
         return notification.type
     }
+    
+    func updateWith(editMode: Bool) {
```
@staykids this was reworked slightly with the refactor, but the main part has been renamed to `updateDisplayState`.

tonisevener commented on Wikipedia/Code/NotificationsCenterModelController.swift:
```diff
@@ -0,0 +1,80 @@
+    }
+    
+    var fetchOffset: Int {
+        return cellViewModels.count
```
@staykids yeah this is annoying because Core Data doesn't play great with Swift, but the `key` attribute in the data model is set to non-optional, so Core Data shouldn't allow any records saved with that missing. Unfortunately the class still generates that property as an optional, but I don't think it will be possible to actually have this `key` value be nil.

I do still want to do further testing on this to make sure I'm thinking of the fetch offset right, particularly if we insert a new record at the top from a refresh, then fetch the next page. I think `DiffableDataSource` should hide any duplicate cells automatically, but I want to make sure we aren't skipping any records in the database in some instances. I'll report back if I see any bugs.

tonisevener replied:
@staykids I did some light testing on a few things to try and catch any issues with this:

1. After the first page is fetched and displayed, I faked a new refresh that caused cells to inserted at the top of the list, then I scrolled down until it fetches the next page. I compared the new records inserted at the bottom with the records in the sqlite database, and didn't see any records skipped. So I think we're good on that.
2. I changed `fetchOffset` to return `cellViewModels.count - 3` to force the database to return duplicate records that are already displaying. Because `cellViewModels` is a Set, it didn't add any duplicate records and the data source stayed unique. I didn't see any dupes shown after the snapshot updated. I think we're good here too.
3. I tried to insert a `nil` key [here](https://github.com/wikimedia/wikipedia-ios/blob/ee33257b9879699fc74499fc82625b877900d166/WMF%20Framework/Remote%20Notifications/Model/RemoteNotificationsModelController.swift#L197) into the database, but it failed and the sqlite db remained empty.

Hope this helps! Let me know if you think I missed anything here.

staykids replied:
Awesome – thanks for this investigative work!

tonisevener commented on Wikipedia/Code/NotificationsCenterViewController.swift:
```diff
@@ -84,41 +96,93 @@ final class NotificationsCenterViewController: ViewController {
+        dataSource = DataSource(
+        collectionView: notificationsView.collectionView,
+        cellProvider: { [weak self] (collectionView, indexPath, viewModel) ->
+            UICollectionViewCell? in
```
@staykids Done!

tonisevener commented on Wikipedia/Code/NotificationsCenterViewController.swift:
```diff
@@ -10,37 +10,49 @@ final class NotificationsCenterViewController: ViewController {
+        
+        setupCollectionView()
+        setupDataSource()
+        configureEmptyState(isEmpty: true)
```
@staykids Thanks for pointing that out. I think we need to show this importing state in a delayed manner, like after a delay of 1 second if it's still importing, display it. But if data populates quickly don't bother to avoid a flash. I started to explore debouncing with `performSelector` and `NSObject.cancelPreviousPerformRequests` like we do elsewhere in the app but it started to feel really gross. If you don't mind I'd rather tackle it in a later PR (separately or whenever we revisit empty states for the filter status).  For now I commented it out and left a TODO note.

tonisevener commented:
@staykids This is ready for another review!

tonisevener commented:
> Have a sufficiently long list of notifications (more than a screenful)
Set the fetch limit to 5: fetchNotifications(fetchLimit: Int = 5...
Toggle editing mode on and start scrolling the list
Some of the incoming displayed notifications are not displayed in the edit mode state, but are instead displayed in the default display state (their leading image is the notification type instead of the checkmark selection view)

@staykids I couldn't repro this bug, but it could be that the edit mode refactor fixed something with it. Let me know if you're still seeing it - if so maybe we can sync up on it.

tonisevener commented on Wikipedia/Code/NotificationsCenterViewController.swift:
```diff
@@ -184,6 +200,14 @@ extension NotificationsCenterViewController: UICollectionViewDelegate {
         }
     }
+    
+    func collectionView(_ collectionView: UICollectionView, shouldBeginMultipleSelectionInteractionAt indexPath: IndexPath) -> Bool {
```
@staykids just FYI this is probably not a must if it will interfere with your swipe action handling. I just added it to see if we can get any bonus functionality by using the system editing mode.

staykids commented:
> > Some of the incoming displayed notifications are not displayed in the edit mode state, but are instead displayed in the default display state (their leading image is the notification type instead of the checkmark selection view)
> 
> @staykids I couldn't repro this bug, but it could be that the edit mode refactor fixed something with it. Let me know if you're still seeing it - if so maybe we can sync up on it.

@tonisevener  Ahh – like a magician, can I guess your card: is the device/sim you're unable to reproduce it on running iOS 14? Try toggling the editing on and then scrolling on iOS 15 (without any code changes) and lemme know if you notice it 🪄

I think I know what's up. iOS 15 introduces some [automatic cell prefetching behavior changes](https://developer.apple.com/videos/play/wwdc2021/10252/). This means that on iOS 15, `visibleCells` isn't necessarily returning all freshly configured and ready to be displayed cells anymore. This matches the behavior I'm seeing on iOS 15 – cells that were likely prefetched were not configured with the new display state, and when you initially begin scrolling after toggling edit on, there appeared to be a few in the list that were displayed in their original display state.


A `notificationsView.collectionView.reloadData()` in `override func setEditing(_ editing: Bool...)` resolves this for iOS 14 & 15, but I think the preferred API for this problem is [reconfigureItems(...)](https://developer.apple.com/documentation/uikit/nsdiffabledatasourcesnapshot/3804468-reconfigureitems). So maybe something like: 

```swift
if #available(iOS 15.0, *) {
	if var snapshot = dataSource?.snapshot() {
		snapshot.reconfigureItems(snapshot.itemIdentifiers)
		dataSource?.apply(snapshot)
	}
} else {
    notificationsView.collectionView.visibleCells.forEach { cell in
        guard let cell = cell as? NotificationsCenterCell else {
            return
        }
        
        cell.configure(theme: theme, isEditing: editing)
    }
}
```

tonisevener commented:
> is the device/sim you're unable to reproduce it on running iOS 14? 

@staykids You are correct sir, that is absolutely what happened. Thanks for figuring that one out and for finding the fix! I applied your changed and didn't see any problems with my testing.

staykids commented on Wikipedia/Code/NotificationsCenterViewController.swift:
```diff
@@ -70,17 +77,35 @@ final class NotificationsCenterViewController: ViewController {
+        if #available(iOS 15.0, *) {
+            if var snapshot = dataSource?.snapshot() {
+                snapshot.reconfigureItems(snapshot.itemIdentifiers)
+                dataSource?.apply(snapshot)
```
In this case, I think it may be preferable to not animate: `dataSource?.apply(snapshot, animatingDifferences: false)` – it's unexpected seeing the leading image fade between states rather than just discretely change between states.

staykids reviewed (commented):
Nice – these all seems like positive changes. So far, I haven't noticed any regressions from the previous round of review. Few minor comments and a note about  `reloadCellWithViewModelIfNeeded` below.

tonisevener commented:
@staykids thanks! I pushed a few commits (some are oddly hidden in the timeline, so be sure to check the Commits tab) addressing your feedback. I'll look it over again in the morning with fresh eyes, it could still maybe use some cleanup.

staykids reviewed (approved):
👍🏽 – I'll hold off on merging until after the Xcode 13 stuff is resolved

staykids commented:
Just a heads up that a few conflicts have snuck into this one.
